src/linear_algebra/affine_space.lean

```diff
@@ -0,0 +1,180 @@
+/-
+Copyright (c) 2020 Joseph Myers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Joseph Myers.
+-/
+import algebra.add_torsor
+import linear_algebra.basis
+
+noncomputable theory
+
+/-!
+# Affine spaces
+
+This file defines affine spaces and subspaces and affine maps and the
+affine span of a set of points.
+
+## Implementation notes
+
+This file is very minimal and many things are surely omitted. Most
+results can be deduced from corresponding results for vector spaces.
+The variables `k` and `V` are explicit rather than implicit arguments
+to lemmas because otherwise the elaborator sometimes has problems
+inferring appropriate types and type class instances.  Definitions of
+affine spaces vary as to whether a space with no points is permitted;
+here, we require a nonempty type of points (via the definition of
+torsors requiring a nonempty type).
+
+## References
+
+* https://en.wikipedia.org/wiki/Affine_space
+* https://en.wikipedia.org/wiki/Principal_homogeneous_space
+
+-/
+
+/-- `affine_space` is an abbreviation for `add_torsor` in the
+case where the group is a vector space. -/
+@[nolint unused_arguments]
+abbreviation affine_space (k : Type*) (V : Type*) (P : Type*) [field k] [add_comm_group V]
+    [vector_space k V] :=
+add_torsor V P
+
+namespace affine_space
+
+open add_action
+open add_torsor
+
+variables (k : Type*) (V : Type*) {P : Type*} [field k] [add_comm_group V] [vector_space k V]
+          [S : affine_space k V P]
+include S
+
+/-- The vector subspace spanning the differences of a (possibly empty)
+set of points. -/
+def vector_span (s : set P) : subspace k V := submodule.span k (vsub_set V s)
+
+/-- The points in the affine span of a (possibly empty) set of
+points. -/
+def span_points (s : set P) : set P :=
+{p | ∃ p1 ∈ s, ∃ v ∈ (vector_span k V s).carrier, p = v +ᵥ p1}
```

Comment:
You shouldn't need to write `.carrier` (and it will be annoying down the road if you do, because there are no `simp`-lemmas for it). You may need to write `\exists v : P, h : v \in (vector_span k V s)`.

Reply:
Removed all uses of `.carrier` for vector subspaces.